src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs

```diff
@@ -369,23 +369,40 @@ private SelectItem GenerateExpandItem(ExpandTermToken tokenIn)
                 currentNavProp = ParseComplexTypesBeforeNavigation(currentComplexProp, ref firstNonTypeToken, pathSoFar);
             }
 
+            bool isRef = false;
+            bool isCount = false;
+            bool hasDerivedTypeSegment = false;
+            IEdmType derivedType = null;
+
+            // Handle $expand=Customer/Fully.Qualified.Namespace.VipCustomer
+            // The deriveTypeToken is Fully.Qualified.Namespace.VipCustomer
+            if (firstNonTypeToken.NextToken != null && firstNonTypeToken.NextToken.IsNamespaceOrContainerQualified())
+            {
+                hasDerivedTypeSegment = true;
+                derivedType = UriEdmHelpers.FindTypeFromModel(this.Model, firstNonTypeToken.NextToken.Identifier, this.configuration.Resolver);

```

Comment:
>derivedType = UriEdmHelpers.FindTypeFromModel [](http://example.com/codeflow?start=16&length=45)

do you need any special handling for FindTypeFromModel not finding the type? In that case, I believe derivedType would return null, and it looks like CheckRelatedTo may throw a null ref exception as IsRelatedTo will return false, but the exception that we try to generate on line 57 of EdmUriHelpers will try to call FullTypeName() on the null value.

Minimally, we should fix EdmUriHelpers.CheckRelatedTo to handle the null childType and add tests for adding a cast segment with a type that is not found.

Reply:
When `FindTypeFromModel` doesn't find the type it returns a null. I have added code to throw an exception when the `derivedType` is `null`.

Regarding `EdmUriHelpers.CheckRelatedTo`, I have handled the null `childType` similar to how a null `parentType` is handled.